cmdpolling/polling.go

```diff
@@ -1,15 +1,152 @@
 package cmdpolling
 
 import (
+	"context"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
 	"github.com/codegangsta/cli"
+	"github.com/ingrammicro/concerto/api/polling"
+	"github.com/ingrammicro/concerto/cmd"
+	"github.com/ingrammicro/concerto/utils"
+	"github.com/ingrammicro/concerto/utils/format"
+)
+
+const (
+	PollingPingTimingInterval = 2
+	ProcessIdFile             = "imco-polling.pid"
+)
+
+var (
+	commandProcessed = make(chan bool, 1)
 )
 
-func SubCommands() []cli.Command {
-	return []cli.Command{
-		{
-			Name:   "register",
-			Usage:  "Register concerto agent within an imported Host",
-			Action: cmdRegister,
-		},
+// Handle signals
+func handleSysSignals(cancelFunc context.CancelFunc) {
+	log.Debug("handleSysSignals")
+
+	gracefulStop := make(chan os.Signal, 1)
+	signal.Notify(gracefulStop, syscall.SIGTERM, syscall.SIGINT, syscall.SIGKILL)
+	log.Debug("Ending, signal detected:", <-gracefulStop)
+	cancelFunc()
+}
+
+// Start the polling process
+func cmdStart(c *cli.Context) error {
+	log.Debug("cmdStart")
+
+	formatter := format.GetFormatter()
+	if err := utils.SetProcessIdToFile(ProcessIdFile); err != nil {
+		formatter.PrintFatal("cannot create the pid file", err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	go handleSysSignals(cancel)
+
+	pingRoutine(ctx, c)
+
+	return nil
+}
+
+// Stop the polling process
+func cmdStop(c *cli.Context) error {
+	log.Debug("cmdStop")
+
+	formatter := format.GetFormatter()
+	if err := utils.StopProcess(ProcessIdFile); err != nil {
+		formatter.PrintFatal("cannot stop the polling process", err)
 	}
+
+	log.Info("concerto polling successfully stopped")
+	return nil
+}
+
+// Main polling background routine
+func pingRoutine(ctx context.Context, c *cli.Context) {
+	log.Debug("pingRoutine")
+
+	formatter := format.GetFormatter()
+	pollingSvc := cmd.WireUpPolling(c)
+
+	isRunningCommandRoutine := false
+	t := time.NewTicker(PollingPingTimingInterval * time.Second)
+	for {
+		log.Debug("Requesting for candidate commands status")
+		ping, status, err := pollingSvc.Ping()
+		if err != nil {
+			formatter.PrintError("Couldn't receive polling ping data", err)
+		} else {
+			// One command is available, and no process running
+			if status == 201 && ping.PendingCommands && !isRunningCommandRoutine {
+				log.Debug("Detected a candidate command")
+				isRunningCommandRoutine = true
+				go processingCommandRoutine(pollingSvc, formatter)
+			}
+		}
+
+		select {
+		case <-commandProcessed:
+			isRunningCommandRoutine = false
+		case <-ctx.Done():
+			log.Debug(ctx.Err())
+			log.Debug("closing polling")
+			return
+		default:
+		}
+		<-t.C
```

Comment:
The problem with this is that the process may take up to PollingPingTimingInterval seconds + <time it takes to make another ping request> to process a signal, when the process could handle it during the sleep. In order to do that, you should keep the select above and replace this line with another select:
```Go
select {
  case <- t.C:
  case <- ctx.Done():
    // TODO: log debug messages
     return
}
```